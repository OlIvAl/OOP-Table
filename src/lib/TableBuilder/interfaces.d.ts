/**
 * Колонка в компоненте ReactTable
 */
export interface IColumn<R> {
  /**
   * Ключ доступа к колонке ы таблице
   */
  accessor: keyof R;
  /**
   * Класс колонки таблицы
   */
  headerClassName: keyof R;
}

/**
 * Данные для компонента ReactTable
 * @param R Объект строки
 */
export interface ITable<R> {
  /**
   * Map обЪектов колонок
   */
  columns: Record<keyof R, IColumn<R>> | null;
  /**
   * Массив обЪектов строк
   */
  rows: R[];
  footerRow: Record<keyof R, string> | null;
}

/**
 * функция для преобразования моделей бизнесс данных в объекты строк
 * @param M Объект модели бизнесс данных
 * @param R Объект строки
 */
type IModelsToRowsConverter<M, R> = (model: M) => R;
/*
/!**
 * фабрика для преобразования моделей бизнесс данных в объекты строк
 * @param M Объект модели бизнесс данных
 * @param R Объект строки
 *!/
interface IModelsToRowsFabric<M, R> {
  getRow: () => R;
}
*/

export interface IHandler<R> {
  setNext(handler: IHandler): IHandler;

  handle(request: ITable<R>): ITable<R>;
}

/**
 * Универсальный конструктор таблиц
 * @param M Объект модели бизнесс данных
 * @param R Объект строки
 */
export interface ITableBuilder<M, R> {
  /**
   * Сетит ключи колонок для дальнейшего построения объектов колонок
   * @param {string[]} columnKeys ключи колонок
   */
  setColumnKeys: (columnKeys: string[]) => void;
  /**
   * Сетит бизнесс модели для дальнейшего преобразования их в строки
   * @param models модели для дальнейшего преобразования их в строки
   */
  setModels: (models: M[]) => void;
  /**
   * Сетит функцию для преобразования моделей бизнесс данных в объекты строк
   * @param modelsToRowsConverter функция для преобразования моделей бизнесс данных в объекты строк
   */
  setModelsToRowsConverter: (modelsToRowsConverter: IModelsToRowsConverter<M, R>) => void;
  /**
   * Сетит данные для строки футера
   * @param footerRow данные для строки футера
   */
  setFooter: (footerRow: Record<keyof R, string> | null) => void;

  /**
   * Возвращает готовый для использования в компоненте ReactTable объект таблицы
   * @returns готовый для использования в компоненте ReactTable объект таблицы
   */
  getTable: () => ITable<R>;
  /**
   * Ресет начальных значений
   */
  reset: () => void;
}

/**
 * Директор для формирования стандартных таблиц
 * @param R Объект строки
 */
export interface ITableDirector<R> {
  /**
   * Метод формирует простую тадлицу из переданных строк
   * Строки формируются из ключей объекта строки
   * @param tableBuilder объект билдера таблиц
   * @param rows массив строк
   */
  buildSimpleTable: (tableBuilder: ITableBuilder<R, R>, rows: R[]) => void;
  /**
   * Метод формирует простую тадлицу из переданных строк и строки футера
   * Строки формируются из ключей объекта строки
   * @param tableBuilder объект билдера таблиц
   * @param rows массив строк
   * @param footerRow строка футера
   */
  buildSimpleTableWithFooter: (
    tableBuilder: ITableBuilder<R, R>,
    rows: R[],
    footerRow: Record<keyof R, string> | null
  ) => void;
}
