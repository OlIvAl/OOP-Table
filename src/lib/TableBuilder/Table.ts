import { IColumn, ITable } from './interfaces';

export class Table<R> implements ITable<R> {
  columns: Record<keyof R, IColumn<R>> | null = null;
  rows: R[] = [];
  footerRow: Record<keyof R, string> | null = null;
}
