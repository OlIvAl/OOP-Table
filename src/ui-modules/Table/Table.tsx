import React from 'react';
import { TableOptions, useTable } from 'react-table';

interface IProps extends TableOptions<any> {
  loading: boolean
}

export const Table: React.FC<IProps> = ({ columns, data , loading}): JSX.Element => {
  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    rows,
    prepareRow,
  } = useTable({
    columns,
    data,
  });

  if( loading ) {
    return <div>Loading...</div>;
  }

  return (
    <table {...getTableProps()}>
      <thead>
      {headerGroups.map(headerGroup => (
        <tr {...headerGroup.getHeaderGroupProps()}>
          {headerGroup.headers.map(column => (
            <th {...column.getHeaderProps()}>{column.render('Header')}</th>
          ))}
        </tr>
      ))}
      </thead>
      <tbody {...getTableBodyProps()}>
      {rows.map(
        (row) => {
          prepareRow(row);
          return (
            <tr {...row.getRowProps()}>
              {row.cells.map(cell => {
                return <td {...cell.getCellProps()}>{cell.render('Cell')}</td>
              })}
            </tr>
          )}
      )}
      </tbody>
    </table>
  );
};