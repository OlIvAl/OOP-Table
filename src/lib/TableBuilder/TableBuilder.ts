import { IModelsToRowsConverter, ITable, ITableBuilder } from './interfaces';
import { Table } from './Table';
import { InitColumnsHandler } from './tableHandlers/InitColumnsHandler';
import { InitRowsHandler } from './tableHandlers/InitRowsHandler';
import { InitFooterRowHandler } from './tableHandlers/InitFooterRowHandler';

export class TableBuilder<M, R = M> implements ITableBuilder<M, R> {
  /**
   * готовый для использования в компоненте ReactTable объект таблицы
   */
  private table: ITable<R>;
  /**
   * ключи колонок
   * @default []
   */
  private columnKeys: (keyof R)[];
  /**
   * модели для дальнейшего преобразования их в строки
   * @default []
   */
  private models: M[];
  /**
   * функция для преобразования моделей бизнесс данных в объекты строк
   * @default null
   */
  private modelsToRowsConverter: IModelsToRowsConverter<M, R> | null;
  /**
   * Объект строки футера
   * @default null
   */
  private footerRow: Record<keyof R, string> | null;

  constructor() {
    this.table = new Table();
    this.columnKeys = [];
    this.models = [];
    this.modelsToRowsConverter = null;
    this.footerRow = null;

    this.reset = this.reset.bind(this);
    this.setColumnKeys = this.setColumnKeys.bind(this);
    this.setModels = this.setModels.bind(this);
    this.setModelsToRowsConverter = this.setModelsToRowsConverter.bind(this);
    this.setFooter = this.setFooter.bind(this);
    this.getTable = this.getTable.bind(this);
  }

  reset(): void {
    this.table = new Table();
    this.columnKeys = [];
    this.models = [];
    this.modelsToRowsConverter = null;
    this.footerRow = null;
  }

  setColumnKeys(columnKeys: string[]): void {
    if (!columnKeys.length) {
      throw Error('You must to set array of column keys. You set empty array.');
    }
    this.columnKeys = columnKeys as (keyof R)[];
  }

  setModels(models: M[]): void {
    this.models = models;
  }

  setModelsToRowsConverter(modelsToRowsConverter: IModelsToRowsConverter<M, R>): void {
    this.modelsToRowsConverter = modelsToRowsConverter;
  }

  setFooter(footerRow: Record<keyof R, string> | null): void {
    this.footerRow = footerRow;
  }

  getTable(): ITable<R> {
    if (!this.columnKeys.length) {
      throw Error('You have not set column keys');
    }

    const initColumnsHandler = new InitColumnsHandler<M, R>(this.columnKeys);
    const initRowsHandler = new InitRowsHandler<M, R>(this.models, this.modelsToRowsConverter);
    const initFooterRowHandler = new InitFooterRowHandler<M, R>(this.footerRow);

    initColumnsHandler.setNext(initRowsHandler).setNext(initFooterRowHandler);

    return initColumnsHandler.handle(this.table);
  }
}
