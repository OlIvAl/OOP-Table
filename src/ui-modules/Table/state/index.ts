import { action, computed, flow, IReactionDisposer, reaction } from 'mobx';
import { BalanceRowDecorator } from './decorators/BalanceRowDecorator';
import { BalanceTotalRowDecorator } from './decorators/BalanceTotalRowDecorator';
import { ITable } from '../../../lib/TableBuilder/interfaces';
import { IFinanceActionModel } from '../../../store/Actions/interfaces';
import { TableBuilder } from '../../../lib/TableBuilder/TableBuilder';
import { IBalanceParams, IFinanceActionData } from '../../../DataLayer/modules/actions/interfaces';
import { IStoreWithCurrentModel } from '../../../store/Abstract/interfaces';
import { IBalanceRow, IBalanceTotalRow, IFinanceActionsDataService } from './interfaces';
import { FinanceActionsStoreInitDecorator } from './decorators/FinanceActionsStoreInitDecorator';

/**
 * Объект состояния компонента баланса
 */
export class BalanceTableState {
  @computed get loading(): boolean {
    return this.getFinanceActionsService.loading;
  }
  @computed get balanceTableShown(): boolean {
    return !!this.financeActionsStore.models.length;
  }
  @computed get table(): ITable<IBalanceRow> | null {
    const models: IFinanceActionModel[] = this.financeActionsStore.models;
    if (!models.length) {
      return null;
    }

    const rows: IBalanceRow[] = models.map<IBalanceRow>(model => new BalanceRowDecorator(model));

    const balanceTableBuilder = new TableBuilder<IBalanceRow>();
    const totalRow: IBalanceTotalRow | null = this.getFinanceActionsService.total
      ? new BalanceTotalRowDecorator(this.getFinanceActionsService.total, this.currency)
      : null;

    balanceTableBuilder.setColumnKeys([
      'source',
      'created_at',
      'updated_at',
      'description',
      'action_state',
      'price',
      'profit_diff',
      'total_profit'
    ]);
    balanceTableBuilder.setModels(rows);
    balanceTableBuilder.setFooter(totalRow);

    return balanceTableBuilder.getTable();
  }

  protected financeActionsStore: IStoreWithCurrentModel<IFinanceActionModel>;
  protected getFinanceActionsService: IFinanceActionsDataService;
  protected financeActionsDataHandler: IReactionDisposer;

  protected get currency(): string {
    return 'rub';
  }

  constructor(
    financeActionsStore: IStoreWithCurrentModel<IFinanceActionModel>,
    getFinanceActionsService: IFinanceActionsDataService
  ) {
    this.financeActionsStore = financeActionsStore;
    this.getFinanceActionsService = getFinanceActionsService;

    this.financeActionsDataHandler = reaction<IFinanceActionData[]>(
      () => this.getFinanceActionsService.actions,
      (actions): void => {
        new FinanceActionsStoreInitDecorator(this.financeActionsStore).init(actions);
      }
    );
  }

  @action.bound async init(): Promise<void> {
    const params: IBalanceParams = {};

    await flow(this.getFinanceActionsService.init)(params);
  }

  /**
   * Метод получения списка букингов по передаваемым параметрам
   * @param params Объект с данными для запроса букингов
   */
  @action.bound async getFinanceActions(params: IBalanceParams): Promise<void> {
    await flow(this.getFinanceActionsService.init)(params);
  }

  /**
   * Сбрасывает хранилище
   */
  @action.bound dispose(): void {
    this.financeActionsStore.resetStore();
    this.financeActionsDataHandler();
  }
}
