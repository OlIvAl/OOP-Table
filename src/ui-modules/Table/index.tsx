import React from 'react';
import { Table } from './Table';
import { observable } from 'mobx';
import { RootStore } from '../../store/RootStore';
import { StoreContext } from '../../index';
import { BalanceTableState } from './state';
import { observer } from 'mobx-react-lite';
import { FinanceActionsDataService } from './services/FinanceActionsDataService';

const BalanceTable: React.FC = (): JSX.Element | null => {
  const getFinanceActionsService = new FinanceActionsDataService();
  const { financeActionsStore } = React.useContext<RootStore>(StoreContext);
  const [balanceTableState] = React.useState<BalanceTableState>(() =>
    observable(new BalanceTableState(financeActionsStore, getFinanceActionsService))
  );

  React.useEffect(() => {
    balanceTableState.init();

    return balanceTableState.dispose;
  }, [balanceTableState]);

  const table = balanceTableState.table;

  if (!table || !table.columns) {
    return null;
  }

  const columns = [
    {
      Header: () => 'Source',
      headerClassName: table.columns.source.headerClassName,
      accessor: table.columns.source.accessor
    },
    {
      Header: () => 'Created at',
      accessor: table.columns.created_at.accessor,
      headerClassName: table.columns.created_at.headerClassName
    },
    {
      Header: () => 'Updated at',
      accessor: table.columns.updated_at.accessor,
      headerClassName: table.columns.updated_at.headerClassName
    },
    {
      Header: () => 'Description',
      headerClassName: table.columns.description.headerClassName,
      accessor: table.columns.description.accessor
    },
    {
      Header: () => 'Price',
      headerClassName: table.columns.price.headerClassName,
      accessor: table.columns.price.accessor
    },
    {
      Header: () => 'Total profit',
      headerClassName: table.columns.total_profit.headerClassName,
      accessor: table.columns.total_profit.accessor
    }
  ];

  return (
    <section className="balance_actions_table">
      <Table
        data={table.rows}
        columns={columns}
        loading={balanceTableState.loading}
      />
    </section>
  );
};

export default observer(BalanceTable);