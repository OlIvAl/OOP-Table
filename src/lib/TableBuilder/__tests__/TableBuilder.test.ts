import { ITableBuilder } from '../interfaces';
import { TableBuilder } from '../TableBuilder';
import { Table } from '../Table';
import {
  columnKeys,
  columns,
  getFakeModels,
  getFooterRow,
  ITestModel,
  ITestRow,
  modelsToRowsConverter
} from '../__fixtures__/tableData';

let tableBuilder: ITableBuilder<ITestModel, ITestRow>;
let tableBuilderForDecorators: ITableBuilder<ITestRow, ITestRow>;
let models: ITestModel[];

let rows: ITestRow[];
let footerRow: ITestRow;
const table = new Table<ITestRow>();
const tableWithFooter = new Table<ITestRow>();

beforeEach(() => {
  tableBuilder = new TableBuilder<ITestModel, ITestRow>();
  tableBuilderForDecorators = new TableBuilder<ITestRow, ITestRow>();
  models = getFakeModels();

  footerRow = getFooterRow();

  rows = models.map(modelsToRowsConverter);

  table.rows = rows;
  table.columns = columns;

  tableWithFooter.rows = rows;
  tableWithFooter.columns = columns;
  tableWithFooter.footerRow = footerRow;
});

describe('TableBuilder', () => {
  describe('constructor', () => {
    it('init table builder with valid object', () => {
      expect(new TableBuilder<ITestRow, ITestRow>()).toMatchSnapshot();
    });
  });
  describe('setColumnKeys', () => {
    it('call once', () => {
      const spy = jest.spyOn(tableBuilder, 'setColumnKeys');
      tableBuilder.setColumnKeys(columnKeys);

      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(columnKeys);
    });
    it('call with empty array', () => {
      expect(() => tableBuilder.setColumnKeys([])).toThrowError(/.+/);
    });
  });
  describe('setModels', () => {
    it('call once', () => {
      const spy = jest.spyOn(tableBuilder, 'setModels');
      tableBuilder.setModels(models);

      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(models);
    });
  });
  describe('setModelsToRowsConverter', () => {
    it('call once', () => {
      const spy = jest.spyOn(tableBuilder, 'setModelsToRowsConverter');
      tableBuilder.setModelsToRowsConverter(modelsToRowsConverter);

      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(modelsToRowsConverter);
    });
  });
  describe('setFooter', () => {
    it('call once', () => {
      const spy = jest.spyOn(tableBuilder, 'setFooter');
      tableBuilder.setFooter(footerRow);

      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(footerRow);
    });
  });
  describe('getTable', () => {
    it('if you correct init builder, it return valid table object', () => {
      tableBuilder.setColumnKeys(columnKeys);
      tableBuilder.setModels(models);
      tableBuilder.setModelsToRowsConverter(modelsToRowsConverter);

      const currentTable = tableBuilder.getTable();

      expect(currentTable).toEqual(table);
    });
    it('you can not pass converter and pass ready rows, it return valid table object', () => {
      tableBuilderForDecorators.setColumnKeys(columnKeys);
      tableBuilderForDecorators.setModels(rows);

      const currentTable = tableBuilderForDecorators.getTable();

      expect(currentTable).toEqual(table);
    });
    it('set column keys is required for call getTable', () => {
      tableBuilder.setModels(models);
      tableBuilder.setModelsToRowsConverter(modelsToRowsConverter);

      expect(tableBuilder.getTable).toThrowError(/.+/);
    });
  });
  describe('getTable with footer', () => {
    it('if you correct init builder, it return valid table object', () => {
      tableBuilder.setColumnKeys(columnKeys);
      tableBuilder.setModels(models);
      tableBuilder.setModelsToRowsConverter(modelsToRowsConverter);
      tableBuilder.setFooter(footerRow);

      const currentTable = tableBuilder.getTable();

      expect(currentTable).toEqual(tableWithFooter);
    });
    it('you can not pass converter and pass ready rows, it return valid table object', () => {
      tableBuilderForDecorators.setColumnKeys(columnKeys);
      tableBuilderForDecorators.setModels(rows);
      tableBuilderForDecorators.setFooter(footerRow);

      const currentTable = tableBuilderForDecorators.getTable();

      expect(currentTable).toEqual(tableWithFooter);
    });
  });
  describe('reset', () => {
    it('reset table columnKeys models modelsToRowsConverter footerRow', () => {
      tableBuilder.setColumnKeys(columnKeys);
      tableBuilder.setModels(models);
      tableBuilder.setModelsToRowsConverter(modelsToRowsConverter);
      tableBuilder.setFooter(footerRow);

      const currentTable = tableBuilder.getTable();

      expect(currentTable).toEqual(tableWithFooter);

      tableBuilder.reset();

      //@ts-ignore
      expect(tableBuilder.table.columns).toEqual(null);
      //@ts-ignore
      expect(tableBuilder.columnKeys).toEqual([]);
      //@ts-ignore
      expect(tableBuilder.models).toEqual([]);
      //@ts-ignore
      expect(tableBuilder.modelsToRowsConverter).toEqual(null);
      //@ts-ignore
      expect(tableBuilder.footerRow).toEqual(null);
    });
  });
});
