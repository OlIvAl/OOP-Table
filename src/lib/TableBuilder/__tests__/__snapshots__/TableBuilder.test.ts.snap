// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableBuilder constructor init table builder with valid object 1`] = `
TableBuilder {
  "columnKeys": Array [],
  "footerRow": null,
  "getTable": [Function],
  "models": Array [],
  "modelsToRowsConverter": null,
  "reset": [Function],
  "setColumnKeys": [Function],
  "setFooter": [Function],
  "setModels": [Function],
  "setModelsToRowsConverter": [Function],
  "table": Table {
    "columns": null,
    "footerRow": null,
    "rows": Array [],
  },
}
`;
